compiler/circle-opselector/src/OpSelector.cpp

```diff
@@ -0,0 +1,505 @@
+    std::vector<const circle::OperatorT *>
+      input_nodes; // the nodes that one's input is not connected with other node's output
+    std::vector<const circle::OperatorT *>
+      output_nodes; // the nodes that one's output is not connected with other node's input
+
+    _reader.select_subgraph(g); // select subgraph. usually, 0 is main
+
+    graph->name(_reader.name()); // set name of graph (if it has one graph, name is empty.)
+
+    const auto &operators = _reader.operators(); // operator is the nodes that we can see in netron.
+                                                 // It has node's input, output, etc..
+    const auto &tensors = _reader.tensors();     // tensor is operator's input or output node.
+    auto tensors_ptr = _reader.tensors_ptr();
+    assert(tensors_ptr != nullptr);
+
+    auto nodefinder = std::make_unique<luci::IndexNodeFinder>(); // node find helper
+    auto tensoroutputs = std::make_unique<luci::IndexTensorOutputs>();
+
+    luci::GraphBuilderContext gb_context(graph.get(), &_reader, nodefinder.get(),
+                                         tensoroutputs.get()); // graph build helper.
+
+    // print selected operator's detail.
+    std::cout << "Subgraph Name: " << _reader.name() << std::endl;
+    for (auto cnode : selected_nodes)
+    {
+      uint32_t node_id = luci::get_node_id(cnode);
+      std::string node_name = cnode->name();
+
+      selected_operators.push_back(operators[node_id].get()); // put selected nodes in vector.
+      std::cout << "============== Operator Name: " << node_name << " ==============" << std::endl;
+      std::cout << "    <INPUT>" << std::endl;
+      for (auto node : operators[node_id].get()->inputs)
+        std::cout << "operator[" << node_id << "] id: " << node << " "
+                  << "input: " << tensors[node]->name << std::endl;
+      std::cout << "    <OUTPUT>" << std::endl;
+      for (auto node : operators[node_id].get()->outputs)
+        std::cout << "operator[" << node_id << "] id: " << node << " "
+                  << "output: " << tensors[node]->name << std::endl;
+      std::cout << std::endl;
+    }
+
+    // find the node that has no preceding node
+    input_nodes.push_back(selected_operators[0]); // first node must not have preceding node.
+    for (uint32_t node1 = 1; node1 < selected_operators.size(); ++node1)
+    {
+      bool input_connected = false;
+
+      for (uint32_t node2 = 0; node2 < selected_operators.size(); ++node2)
+      {
+        if (node1 == node2)
+          continue;
+        // find input
+        for (auto input : selected_operators[node1]->inputs)
+        {
+          for (auto output : selected_operators[node2]->outputs)
+          {
+            if (input == output)
+              input_connected = true;
+          }
+        }
+      }
+
+      if (!input_connected)
+        input_nodes.push_back(selected_operators[node1]);
+    }
```

Comment:
I don't think we need to run two 4 level nested loops to identify input and output.

We can gather all input nodes and output nodes by iterating `selected_operators`.

Basically, in pseudo code, (note that it is not complete code.)

```c++
for (auto op : selected_operators) {
  for (auto input : op->inputs)
     input_nodes_set += input
  for (auto output: op->outputs)
     output_nodes_set += output
}
graph_input = input_nodes_set - output_nodes_set
graph_output = output_nodes_set - input_node_set
```

For more complete code, please refer to https://github.com/Samsung/ONE/blob/d9e5cde7b79caf8398d75017ef9caad7e3cd62d2/tools/tflitefile_tool/select_operator.py#L1347-L1384